### ISSUE TITLE
MVP for new UI

### ISSUE THREAD

**Issue by AshotN** [ampache/ampache#2360]
As some may know I have been working on a new UI for Ampache. The [React based client](https://github.com/ampache/ampache/tree/react-client). 

The Reborn interface is getting old and slow, it loads a lot of dependencies and is not very intuitive to use. While fixing is it something I considered at first, it's code base is huge and rather messy. To no fault of anyone, but as Ampache is an old project, it has a lot of tacked on features that just simply were left to rot. Fixing the Reborn interface seems to me to be harder than actually starting from scratch. 

This issue can be used to document required features for including the UI as part of the develop branch, and then later we can iron it out into an actual release.

Feature parity with the Reborn interface is obviously a goal, but this is a good opportunity to trim out features that are no longer being maintained or are relevant. 

I'd appreciate it if we can create a list of required features for a minimum viable product. We can also include any features that would be nice to have. 

One thing I am striving for that the Reborn interface doesn't do, mobile support. Using Ampache on a phone browser is not a pleasant experience. While it's always ideal to have a native app, I think having a usable browser experience is still a requirement.

Another aspect is the limitation of the current API, while it's made tremendous progress, it's not powerful enough currently to cover all of the Reborn features.

**Comment by lachlan-00** [comment]
Don't know if others outside the org can see but we have an API extension project.
https://github.com/orgs/ampache/projects/1

I want the API to be able to provide everything 100% that reborn provides.

Would love to hear from others and quiet achiever's like @stebe (and more!) Who give valuable insights into how they use ampache.

Pinning this issue too.

**Comment by AshotN** [comment]
I'd appreciate input from @wagnered and @Afterster espically, especially as they have been with the project for a long time.

**Comment by wagnered** [comment]
@AshotN,

I wholely agree with you.  But are any features not being maintained because no one is using them, or has no problems been reported?  Unit testing might give an answer this.

As for mobile browser support,  it's a decision between an adaptive or responsive delivery of content.  Or both depending on the display size.

To update the API to full compatability with the reborn interface I suggest this task be next in priority after the "typing" project.  Patching current features at the whim of a user should have lowest priority.  Surely, a member other than @lachlan-00 has the skill to assist in the API upgrade.  As soon as I learned the basic PHP concepts, I submitted both bug fix and feature upgrade PRs and learned much from @afterster's review of them.  I would love to take this on, but do not yet have a computer.  

And while we are bringing Ampache logic into the 21st century,  I would like to suggest also bringing its architecture into the 21st century by adapting it to a well defined framework.

@AshotN this might not be the type of information you are expecting but I thought it important to give my view of the whole project.

**Comment by kuzi-moto** [comment]
Maybe with so many more active members, it's worth thinking about Laravel again? I know a lot of work has gone into Ampache in the last year, but I think at this point the project would benefit greatly from a rewrite, and there are more people around to spread the load.

Otherwise, there is a lot of refactoring to be done on the current code.

**Comment by lachlan-00** [comment]
Here are the api methods that are part of the parity project to do.

## Script API Methods

Methods used for editing individual files and in user scripts

* file_add
* file_clean
* file_verify

## Album API Methods

Get albums and the songs in that album.

* album_edit (edit an album database field)

## Artist API Methods

Get details on artists including albums and songs connected to their id.

* artist_edit (edit an artist database field)

## Search API Methods

Searching for items once you have connected to your server.

* get_similar (get similar last.fm artists, albums, songs, etc)

## Song API Methods

Song methods.

* song_edit (edit a song database field)
* song_delete
* song_organize (fix filename according to catalog)

## User API Methods

General user mthods used for getting information about individual users.

* play_queue (get/save a playlist as the current queue)
* bookmark (add/sve your place for a file to resume later)
* pref_interface (User Config preferences)
* pref_options (User Config preferences)
* pref_playlist (User Config preferences)
* pref_plugins (User Config preferences)
* pref_streaming (User Config preferences)

## Administration API Methods

Advanced methods that make changes to the system.

* user_clear_data (now playing, recent, stats, etc)
* catalogs
* catalog
* catalog_action (extend for other types of action)
* catalog_create
* catalog_edit
* catalog_delete
* acls
* acl
* acl_create
* acl_edit
* acl_delete
* ampache_update (update from git)
* ampache_interface (Server preferences)
* ampache_options (Server preferences)
* ampache_playlist (Server preferences)
* ampache_plugins (Server preferences)
* ampache_streaming (Server preferences)
* ampache_system (Server preferences)
* db_optimize (bin/catalog_update)
* db_clean_art (bin/clean_art_table)

## Control API Methods

Methods used for server backends.

* channel (start/stop/status)

## Streaming Radio API Methods

* radios
* radio
* radio_create
* radio_edit
* radio_delete
* channel_create
* channel_edit
* channel_delete

## Podcast API Methods

* podcasts
* podcast
* podcast_episodes
* podcast_episode
* podcast_create
* podcast_edit
* podcast_delete

## Playlist API Methods

Manage your playlists (including smart lists)

* playlist_import (allow radio/urls)
* playlist_export (allow radio/urls)

## Metadata Action API Methods

Update tags and metadata for files.

* metadata_fields (list the custom metadata fields)
* metadata_field (Get a single metadata field by name)
* metadata_create (create a new metadata field)
* metadata_edit (edit a metadata field)
* metadata_remove (remove a metadata field)

## Share API Methods

* shares
* share
* share_create
* share_edit
* share_delete

## Upload API Methods

Upload files and get details about existing uploads

* uploads
* upload
* upload_create
* upload_delete

**Comment by lachlan-00** [comment]
New methods in develop and documented on the wiki
* shares
* share
* get_similar

**Comment by Afterster** [comment]
A good API is clearly a required step. That should have been achieved through the move to Laravel but if not a current mater then yes an Ampache API extension project as you did is clearly a good move IMOO.

Years ago I created the reborn interface with SUTJael to have something modern compared to what existed without rewriting the complete backend (it was my first involvement on Ampache project, didn't know I was going to change a lot more things at that time) ; even if not fully responsive it was a move forward. A react interface may be a good approach.
I see the move to Laravel as a big resource and motivation killer. I believe we failed because we tried to move all used features of Ampache to Laravel. Too much work. On the other side, not doing so would create something else than Ampache, and this probably already exists (eg. Koel or others) and is not why we all use Ampache today. Difficult move, I would prefer to continue with 'baby steps' which are still useful work like API improvement and React client you are talking about. Personal opinion, you don't have to agree ^^.

**Comment by AshotN** [comment]
Laravel would be great, it's much easier to write an API with it. But Ampache is so large that it's quite the task for anyone.

Perhaps after we decouple the interface from the server, it might make it easier to reattempt a Laravel transition.

**Comment by lachlan-00** [comment]
I'm with @Afterster on this, my 2020 goal is API and then advertising that new API out. It won't matter how old the web interface gets once we get that ability into something that's not limited by the existing code.

I'm going to keep trying to make the external documentation process easier for other people to pick up. The Website will be getting a facelift in that regard as well to try and skip the issues with markdown/github as part of the update process.

Code examples and example / official apps are after that.

**Comment by AshotN** [comment]
So some things that I imagine are needed for a new UI to ship. 

- [ ] Ability to edit Ampache settings
- [ ] Handle radio streams
- [ ] Edit Artist/Album/Song information and art
- [ ] Support Ampache's link sharing

**Comment by lachlan-00** [comment]
Moving into a new branch for API work.

https://github.com/ampache/ampache/tree/api4build5

All those features are aiming to be in for sure. Not worthy of an API 5 but probably 4.1 build 1 by the end of this.

Trying to pick off all the missing subsonic ones first as they're essentially already there.

**Comment by lachlan-00** [comment]
New methods added in the api branch. podcasts need more testing but shares are all done now.
* podcasts
* podcast
* share_create
* share_edit
* share_delete

**Comment by lachlan-00** [comment]
fixed up the podcasts. will document later and merge these changes when tested before moving on to the next group of methods
* podcasts
* podcast
* podcast_episodes
* podcast_episode

**Comment by lachlan-00** [comment]
next methods almost ready for testing. it's a long weekend here so i'll probably not be around much but this has been some solid api progress this week
  * podcast_create: create a podcast
  * podcast_edit: edit an existing podcast
  * podcast_delete: delete an existing podcast
  * update_podcast: sync and download new episodes
  * catalogs: get all the catalogs
  * catalog: get a catalog by id

13 new methods in one week though is pretty good. next week this round will probably merge when i get back to the office.

**Comment by lachlan-00** [comment]
catalog_file added to allow actions on individual file paths:
* add
* remove (delete)
* verify (update from tags) 
* clean (delete from db if removed from disk)

I'll also add an organize/sort function to allow sorting individual files into the catalog pattern. which i'll include while i work on wav file catalog issues from #2377.

it's also given me an idea that documentation needs a clearer access requirement for methods as i've put in a lot more feature and access checking to make sure people aren't deleting things like crazy

**Comment by manuelglez86** [comment]
I think improving the API is a good idea, I created a windows interface focused more on creating playlists, I implemented on client side the playlist edit process (i have created an algorithm because there wasnt a way to edit the content of a playlist) but I think that if the focus would be only the API then Ampache would be more for developers and not for the final user because we will need someone to created a multiplataform client.

As an example, we use windows computers for the people involve in creating the playllists but in the broadcast side we use Mac computers, we download the playlists from the web client and play it on other software (like itunes). So we have adapted amapche for our purposes.

But we have to think also on ampache final users, I know that the person who installs ampache will need a little of computer knowledge but not necesary need know how to program and the API itself doesnt give an usable interface.

**Comment by kuzi-moto** [comment]
I was working on redoing the web interface, haven't had time for it lately though. I think it would be the best thing to include, as you can make the design responsive, and multi-platform by default.

**Comment by manuelglez86** [comment]
Yes the “multiplataform” that the web client gives by default is the reason I chose ampache, but the process of creating playlists was a little slow beacuase we have a big catalog, so thats why I decided to create a windows API cliente for that purpose.

But again not everyone could or want to do that, some people just want to install and use.

The point is that we need to define who will be the ampache “client target”. (I dont know if it is the correct marketing term)

**Comment by kuzi-moto** [comment]
I think Ampache will always be for more tech-savvy users, given that it's not a simple .exe like Plex media server. With some improvements to the Docker image, I hope that will give it more popularity.

**Comment by lachlan-00** [comment]
Playlist tracks can be changed now but not removed yet. (I forgot that. It'll be in the release after this one)

I will always be working on Ampache web, I'm not smart enough to make a new UI but I can extend and maintain what we have for years to come. The API is very important to make Ampache as available as possible beyond just the web interface in ways that I can't do.

I will also be willing to adapt the API to help people like you who need a function in Ampache that the API can't do yet. As it stands the web interface is "stable" so doesn't really require much work.

**Comment by lachlan-00** [comment]
@manuelglez86 after make an issue with the things you would like in the API to help you and I'll add them in ASAP.

**Comment by manuelglez86** [comment]
Thank you @lachlan-00 I wanted to make some insides on the new implementation of the playlist_edit but now i saw that you already have implemented some changes so let me try the new changes so I can see what is missing.

Unfortunatly my API client source code is on the office which is currently close and for now is not a close date for reopen, but as soon as I get back to work I will test the new implementations on the API, maybe it will be when you realease v4.2

Be safe, all of you

**Comment by AshotN** [comment]
I haven't forgotten about Ampache, I just haven't been able to find the time :(

**Comment by manuelglez86** [comment]
@lachlan-00 why you said that playlist tracks cant be removed yet if the API has playlist_remove_song

> Playlist tracks can be changed now but not removed yet. (I forgot that. It'll be in the release after this one)
> 
> I will always be working on Ampache web, I'm not smart enough to make a new UI but I can extend and maintain what we have for years to come. The API is very important to make Ampache as available as possible beyond just the web interface in ways that I can't do.
> 
> I will also be willing to adapt the API to help people like you who need a function in Ampache that the API can't do yet. As it stands the web interface is "stable" so doesn't really require much work.

**Comment by lachlan-00** [comment]
> @lachlan-00 why you said that playlist tracks cant be removed yet if the API has playlist_remove_song

haha, that's good then, in my head i was thinking "playlist_edit doesn't have that so it must not be there" I had a lot on my plate this week.

**Comment by lachlan-00** [comment]
> I haven't forgotten about Ampache, I just haven't been able to find the time :(

I wouldn't worry too much @AshotN there's a lot of cleanup/project work i need to get done before i can do another massive API update again as well. The api project still has so many more things to go in.

(I also spent a lot of time 'at work' inside phpstorm instead of completing other projects.)

**Comment by AshotN** [comment]
It would really be helpful if we have UX designer on the team, I am really bad at designing and it takes more time for me to create a design than it does for me to code it.

**Comment by lachlan-00** [comment]
that's probably the hardest thing to find in free software. but if we can get these bones right there will be more interest in helping out.

we've exponentially increased contributions and progress which has been really good to see.

**Comment by ubuntupunk** [comment]
Can't wait to see a new implementation of Ampache. Please include openstreetmaps and better license management.

**Comment by kuzi-moto** [comment]
@ubuntupunk, what would openstreetmaps be used for? Also, if you have some ideas for how licensing is handled, it would be better to create a new feature request as what you're looking for may be handled more by the back end than the interface.

**Comment by ubuntupunk** [comment]
I guess these are feature requests, and can we not have bedrock type of server install with composer? With Git set up already so that we don't have to blacklist the configs and can have permanent swap outs and CDN arrays, self-hosted not just AWS or Google?

**Comment by kuzi-moto** [comment]
> can we not have bedrock type of server install with composer? I know its not php, but cargo install would do a good deal for me right now.

Since Ampache is written in PHP, composer is the easiest way to install dependencies. To use Cargo, the entire project would have to be re-written in Rust which would not be an easy task.

> Or some code With Git set up already so that we don't have to blacklist the configs

Could you elaborate? Nothing should need to be blacklisted for Ampache to work.

> and can have permanent swap outs and CDN arrays, self-hosted not just AWS or Google?

Also, I'm not entirely sure what you mean here. Ampache doesn't use any CDNs and is entirely self-hosted. It will only call out to Google if you set it up to get stuff like album art.

> Then, there's the other way of thinking which seems to suggest flat file, and minimalism is the way to go?

Ampache is slowly improving, but is only maintained by a small number of active developers. Any suggestions on how to improve are welcome, but likely would take a long time to implement on how big the change is.

> geolocation without having to deal with Gsuite.

There is a google maps plugin. I'm not sure how it works, or if it even works at all though.

**Comment by ubuntupunk** [comment]
The use case I have is a shared host, where we need to offload the data storage. Also currently whenever there is a new version of Ampache, I have to move the entire dir out of the way, install the new server and then replace the configs. Am thinking of a solution to backup the configs to a private version control system most likely git. A more modern system would be lighter, faster, more dynamic and take into account easy storage, configuration and even mesh networking. Google ls very proprietary. Is there a way to write modules for the system, so If I want openstreemaps instead of google maps I can just plug that in?  Have not gotten the google maps feature to work yet and am busy playing with the CSS themes feature.

**Comment by kuzi-moto** [comment]
> Also currently whenever there is a new version of Ampache, I have to move the entire dir out of the way, install the new server and then replace the configs and assets.

Hmm, well if you do install Ampache using Git/Composer then updates would be much more seamless. Though I know you were looking to avoid using Composer. An alternative would be a script that does all the work for you, or if you're able to use Docker, then the Ampache image should work fine.

> Google ls very proprietary. Is there a way to write modules for the system, so If I want openstreemaps instead of google maps I can just plug that in?

I don't have much experience with plugins, but theoretically you could copy the Google maps plugin, and try to adapt the code for use with OSM.

Appreciate the questions, but we've already strayed pretty far from the subject of this thread. Any additional questions or new feature requests would would better in their own issue.

**Comment by AshotN** [comment]
> Can't wait to see a new implementation of Ampache. Please include openstreetmaps and better license management.

You are welcome to beta test it.

**Comment by rmdes** [comment]
> > Can't wait to see a new implementation of Ampache. Please include openstreetmaps and better license management.
> 
> You are welcome to beta test it.

How do I proceed to beta test it ?
I have an Ampache instance running already

**Comment by douglasriper** [comment]
If I can help, I'm willing to participate too.
My instance runs on the Develop version.

**Comment by wagnered** [comment]
I started working on a "better license manager". It incorporates a CC JavaScript  package to precisely select the license, locale and authority (country) that the license is effective under.  Only 1 license per track.

**Comment by AshotN** [comment]
https://github.com/ampache/ampache/tree/react-client This is the branch for it.

Follow the steps to add it to your current install.

I will accept any help, I have been kinda busy recently and have not been able to dedicate time to Ampache sadly :(

But if anyone wants to help, you are welcome to!

**Comment by mitchray** [comment]
> It would really be helpful if we have UX designer on the team

I've been using Ampache for a few years and would love to help out with this

I've recently embarked on a complete tear down and rebuild of the Reborn interface for a mobile-first experience, including redesigned icons

I haven't yet seen the React branch so no idea what the current progress looks like, but I wanted to inform you guys so its on your radar. 

Hoping to have something substantial to share Jan-Feb next year, there are some features I've never used that I have to familiarise myself with (Democratic playlist, Localplay etc...)

**Comment by lachlan-00** [comment]
> > It would really be helpful if we have UX designer on the team
> 
> I've been using Ampache for a few years and would love to help out with this
> 
> I've recently embarked on a complete tear down and rebuild of the Reborn interface for a mobile-first experience, including redesigned icons
> 
> I haven't yet seen the React branch so no idea what the current progress looks like, but I wanted to inform you guys so its on your radar. 
> 
> Hoping to have something substantial to share Jan-Feb next year, there are some features I've never used that I have to familiarise myself with (Democratic playlist, Localplay etc...)

@kuzi-moto

**Comment by kuzi-moto** [comment]
@mitchray, Nice! I was trying to do that about a year ago, and then realized I was spending a lot of time just fixing the HTML and it would almost make more sense to re-write most (all??) of the front-end code.

I have a branch in my fork called responsive-theme if you wanted to see what I have so far. Stripping the CSS from Ampache results in a not very usable interface.

**Comment by mitchray** [comment]
I'll definitely be leaving the coding to the professionals! I'm a designer first and a developer third.

I'm creating an interactive non-code prototype of Ampache by considering how the entire thing would look and behave on mobile, which then has the benefit of leading to a cleaner and more streamlined desktop interface.

**Comment by AshotN** [comment]
If you can make a UI in Photoshop or illustrator, that would greatly help.

**Comment by mitchray** [comment]
I know I said I wouldn't be coding but I've decided contributing in this way would be more beneficial, I've added my current progress to [https://github.com/mitchray/ampache/tree/react-client-revised](https://github.com/mitchray/ampache/tree/react-client-revised).

![Initial pass](https://user-images.githubusercontent.com/5735900/100534121-ce56a980-325f-11eb-8153-c5491e69a879.png)

### Changed

- Overall look and feel
- Artist view (laying foundation for further changes)
- Inline the SVGs to control them with CSS
- Revised main menu items
- More details in the album display

### Added

- Material-UI Rating component, needs reset and favorite buttons
- Custom SVG icons (not all implemented in interface yet)
- Additional buttons for the web player

### Removed

- Temporarily! removed majority of hover & focus styles, media queries
- Web player icons 

### Next up

- Revise menu interactions and toggle media player visibility
- Fixing some CSS regressions I introduced
- Reposition some interface elements

**Comment by DJ-SteffEN** [comment]
First of all hello and I think it's a shame that this discussion on the Reborn interface seems to have died down.

yes, it is old and is/seems often rigid ... I still find it appropriate
yes, it definitely makes sense to think about a completely new UI (keyword phone browser)
yes, the current state is a visual disaster in terms of design
yes, extensions of the API functions in interaction with the UI are necessary

but, the Reborn UI is also flexible, exactly because of its age
but, the effort of a completely new UI is often strongly underestimated
but, many modules of a modern UI have already been integrated successfully
but, the API functions have been improved a lot lately

therefore, I started to improve the UI design with the version "5.4.0-release
so, I added two more designs to the Reborn interface (Gray & Balance)
so, I started to adapt the modern elements to today's look (Light Sidebar)
therefore, I started to complete/expand important functions (e.g. search in header)

Questions, about the URL: [https://inthemix.at]
how do you find?, the overall picture of the webpage in the temporal context of the other streaming portals
how do you find?, the concept of open access including a "monitored/automated" login
how do you find?, the design adjustments of the lists of header search and webplayer, as well as the webplayer itself
how do you find?, my suggestion to maintain the Reborn interface in parallel for the time being

**Comment by AshotN** [comment]
Small update in case anyone cares about this, I have picked this back up for now. At least until I setup a solid foundation and fix all the current bugs and performance issues. The changes I have made so far have already made the UI significantly more performant. I'm not sure if I can commit to getting it to where I want it, but hopefully if I set the foundation someone can take the helm.

**Comment by lachlan-00** [comment]
Updated UI has claimed a lot of people over the years, closing unless it's needed again.
